### PR TITLE
Upgrade fontawesome version to latest (v6) from v5 Resolves #2242

### LINKED ISSUE #2242: Upgrade fontawesome version to latest (v6) from v5
### Please confirm that you have searched existing issues in the repo

Yes, I have searched the existing issues

### Any related issues?

_No response_

### What is the area that this feature belongs to?

Author Usability

### Is your feature request related to a problem? Please describe.

Currently Markbind is using only v5 of fontawesome. 
v6 of fontawesome added quite a number of new features which could be useful to users. 

### Describe the solution you'd like

Upgrade fontawesome from v5 to v6

Instructions here: 
https://fontawesome.com/docs/web/setup/upgrade/

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by SPWwj**
**What is the purpose of this pull request?**

- [X] Documentation update
- [ ] Bug fix
- [X] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

**Overview of changes:**
Resolves #2242
Upgrade from fontawesome-free@5.15.4 to fontawesome-free@6.4.0.
The current icons will be preserved. In addition, this PR allows the use of FontAwesome v6 icons
For detail refer to [backward-compatibility](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#backward-compatibility)
Code:
```
:fa-solid-house:
:far-file-code:
:fa-brands-facebook:
:fab-facebook:

```
Output:
image

**Anything you'd like to highlight/discuss:**
V6 has moved regular icons to the pro version.
However, the free icons in v5 should still be usable due to backward compatibility.
Thus user does not need to do anything and no cost in keeping current icons.

The fas and fab icons have been renamed to fa-solid and fa-brands, respectively.

**Testing instructions:**

Create a new project and try out the instructions provided in the [Markbind Icon documentation](https://deploy-preview-2326--markbind-master.netlify.app/userguide/formattingcontents#icons),
with new [v6 icons](https://fontawesome.com/search?o=r&m=free)

**Proposed commit message: (wrap lines at 72 characters)**
Upgrade font awesome version to latest (v6) from v5

---

**Checklist:** :ballot_box_with_check:

- [X] Updated the documentation for feature additions and enhancements
- [X] Added tests for bug fixes or features
- [X] Linked all related issues
- [X] No unrelated changes 

## Migration Guide
We upgraded from fontawesome-free@5.15.4 to fontawesome-free@6.4.0.
The current icons will be preserved (i.e. you don't have to make any changes if you don't want to) due to [Font Awesome's backward compatibility](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#backward-compatibility).
With Version 6
- new icons are available
- some icons have been renamed
- syntax/usage has changed slightly

For details please refer to [Font Awesome's changelog](https://fontawesome.com/docs/changelog/)

### CODE REVIEW

**Comment by damithc** [comment]
> The current icons will be preserved. In addition, this PR allows the use of FontAwesome v6 icons

@SPWwj This is great. In fact, this reduces the need for this to be included in V5 major release, as it will no longer be a breaking change.

File: docs/userGuide/syntax/icons.md [hunk 1]
```diff
@@ -8,19 +8,18 @@ MarkBind supports using Font Icons provided by Font Awesome, Glyphicons and GitH
...
+* font icons: <span style="color: purple">Don't judge the :fa-solid-book: by its cover! {{ icon_dislike }}</span>
 </span>
 </include>
 
 ###### Using Font Awesome Icons
 1. Decide which icon you want to use from the [list of available icons](https://fontawesome.com/icons?d=gallery&m=free).
 1. Construct the MarkBind name for the selected icon by adding the _type prefix_.
    Note: Font Awesome has three different styles for their icons, each with their own type prefix. Here is an example from each type:
-   * _Solid_ (prefix: `fas-`) e.g., :fas-file-code: (actual name `file-code`, MarkBind name `fas-file-code`)
-   * _Regular_ (prefix: `far-`) e.g., :far-file-code: (actual name `file-code`, MarkBind name `far-file-code`)
```
Comment (tlylt) [inline docs/userGuide/syntax/icons.md:20]: What happens to regular? line 18 still says 3 types.

File: docs/userGuide/syntax/icons.md [hunk 2]
```diff
@@ -8,19 +8,18 @@ MarkBind supports using Font Icons provided by Font Awesome, Glyphicons and GitH
...
-   * _Brands_ (prefix: `fab-`): e.g., :fab-github-alt: (actual name `github-alt`, MarkBind name `fab-github-alt`)
+   * _Solid_ (prefix: `fa-solid-`) e.g., :fa-solid-file-code: (actual name `file-code`, MarkBind name `fa-solid-file-code`)
+   * _Brands_ (prefix: `fa-brands-`): e.g., :fa-brands-github-alt: (actual name `github-alt`, MarkBind name `fa-brands-github-alt`)
 
-1. Insert MarkBind name for the icon enclosed within colons to get the icon in your page.<br>
+2. Insert MarkBind name for the icon enclosed within colons to get the icon in your page.<br>
```
Comment (tlylt) [inline docs/userGuide/syntax/icons.md:22]:
```suggestion
1. Insert MarkBind name for the icon enclosed within colons to get the icon in your page.<br>
```

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts [hunk 3]
```diff
@@ -1,65 +1,97 @@
-import octicons, { IconName } from '@primer/octicons';
-import cheerio from 'cheerio';
-
+const octicons = require('@primer/octicons');
+const cheerio = require('cheerio');
```
Comment (tlylt) [inline packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts:2]: any reason why the import syntax is changed here?

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts [hunk 4]
```diff
@@ -1,65 +1,97 @@
...
 const markdownItRegExp = require('markdown-it-regexp');
 
+const ICON_REGEXP = new RegExp(
+  ':(fa[brs]|fa-brands|fa-solid|glyphicon|octicon|octiconlight|mi[forst])-'
+  + '([a-z-]+)~?([a-z-]+)?:',
+);
+
 function getOcticonIcon(iconName: string) {
-  return octicons[iconName as IconName] ?? null;
+  return octicons[iconName] ?? null;
```
Comment (tlylt) [inline packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts:11]: Seemed like an unrelated change to me, why is this code updated to remove `as IconName`?

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts [hunk 5]
```diff
@@ -1,65 +1,97 @@
...
-        materialIconsClass += '-sharp';
-        break;
-      case 'mit':
-        materialIconsClass += '-two-tone';
-        break;
-      default:
-        // .material-icons generates 'Filled' style icons; hence, no suffix is needed for 'mif'.
-      }
-      // Use .align-middle by default to vertically-align the icon with its surrounding text (if any).
-      // Also, replace dashes (-) with underscores (_) to format the icon name properly.
```
Comment (tlylt) [inline packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts:59]: Are these comments no longer relevant? If not we should keep them for context

**Review by tlylt — CHANGES_REQUESTED** [review]
Thanks @SPWwj, some nits. Also, could you fill in the PR description to aid PR review?
Specifically:
- any problem encountered during the migration? any things to note for this migration?
- how to test your change
- from users' POV: what update do they need and what new icons are now available?
- what is the cost of keeping the current icons, if any?
- etc

File: docs/userGuide/syntax/icons.md (hunk 1 above)
Comment (SPWwj) [reply docs/userGuide/syntax/icons.md:20]: v6 has moved the regular icons to the pro version:
image

However, due to the icon aliasing in v6, we can still use v5 regular icons:
```
:fa-solid-house:
:far-file-code:
:fa-brands-facebook:
:fab-facebook:
```
image

image

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts (hunk 3 above)
Comment (SPWwj) [reply packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts]: ESLint does not allow mix of CommonJS module syntax (module.exports = 'foo' or exports.bar = 'hi') with ECMAScript 6 (ES6) new syntax for modules, using import and export keywords 
```
import foo from './foo';
export function bar() { /* ... */ };
```

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts (hunk 5 above)
Comment (SPWwj) [reply packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts:59]: This is still relevant, I have revert them back.

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts (hunk 4 above)
Comment (SPWwj) [reply packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts]: The explanation is similar to the one above regarding the mixing of CommonJS module syntax and ECMAScript 6 (ES6). However, this is indeed an unrelated refactoring of code and should be in a separate PR. The purpose of all this refactoring is to allow other modules to reuse this icon function without changing the current implementation of markdown-it-icons.ts.

**Comment by SPWwj** [comment]
> Thanks @SPWwj, some nits. Also, could you fill in the PR description to aid PR review? Specifically:
> 
> * any problem encountered during the migration? any things to note for this migration?
> * how to test your change
> * from users' POV: what update do they need and what new icons are now available?
> * what is the cost of keeping the current icons, if any?
> * etc

PR description updated. There should be no impact from the users' POV based on FontAwesome's [backward-compatibility](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#backward-compatibility).

File: docs/userGuide/syntax/icons.md (hunk 1 above)
Comment (tlylt) [reply docs/userGuide/syntax/icons.md:20]: We should mention this in our docs/release notes. Could you help add in/draft such relevant info?
If a user who's currently using font-awesome icons and look at the docs
- should be able to find out what significant changes are made, i.e. syntax has changed, some icons have been renamed although the old ones still work
- be encouraged to update their naming to avoid breaking changes in future updates

**Comment by tlylt** [comment]
> > Thanks @SPWwj, some nits. Also, could you fill in the PR description to aid PR review? Specifically:
> > 
> > * any problem encountered during the migration? any things to note for this migration?
> > * how to test your change
> > * from users' POV: what update do they need and what new icons are now available?
> > * what is the cost of keeping the current icons, if any?
> > * etc
> 
> PR description updated. There should be no impact from the users' POV based on FontAwesome's [backward-compatibility](https://fontawesome.com/docs/web/setup/upgrade/whats-changed#backward-compatibility).

Please fill in the proposed commit message as well. Yes I agree that from users' POV there's no change, but I think we should gently mention the differences brought about in this version upgrade (e,g, how it's still backward-compatible, but we may not be able to guarantee that in the future).

File: packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts [hunk 6]
```diff
@@ -7,8 +7,13 @@ function getOcticonIcon(iconName: string) {
   return octicons[iconName as IconName] ?? null;
 }
 
+const ICON_REGEXP = new RegExp(
```
Comment (tlylt) [inline packages/core/src/lib/markdown-it/plugins/markdown-it-icons.ts:10]: We should be using literal as per https://github.com/MarkBind/markbind/pull/1956
...but strange that eslint is not calling this out

File: packages/cli/test/functional/test_site/expected/markbind/material-icons/material-icons-outlined.woff
Comment (tlylt) [inline packages/cli/test/functional/test_site/expected/markbind/material-icons/material-icons-outlined.woff:1]: Are updates to material-icons intended?

Comment (tlylt) [reply packages/cli/test/functional/test_site/expected/markbind/material-icons/material-icons-outlined.woff:1]: hmm seems like these icons should be updated in some earlier PRs but we didn't, hence it's ok to update them here.

**Review by tlylt — APPROVED** [review]
LGTM, thank you @SPWwj 🚀

**Comment by tlylt** [comment]
@all-contributors please add @SPWwj for doc, code
